src/batcher/solve_pair.rs

```diff
@@ -1,16 +1,439 @@
+    fn match_compare(&self, other: &Order) -> Option<OrderPairType> {
+        if !self.attracts(other) {
+            return None;
+        }
+
+        if self.buy_amount <= other.sell_amount && self.sell_amount <= other.buy_amount {
+            Some(OrderPairType::LhsFullyFilled)
+        } else if self.buy_amount >= other.sell_amount && self.sell_amount >= other.buy_amount {
+            Some(OrderPairType::RhsFullyFilled)
+        } else {
+            Some(OrderPairType::BothFullyFilled)
+        }
+    }
+
+    fn attracts(&self, other: &Order) -> bool {
+        self.opposite_tokens(other) && self.have_price_overlap(other)
+    }
+
+    fn opposite_tokens(&self, other: &Order) -> bool {
+        self.buy_token == other.sell_token && self.sell_token == other.buy_token
+    }
+
+    fn have_price_overlap(&self, other: &Order) -> bool {
+        self.sell_amount > U256::zero()
+            && other.sell_amount > U256::zero()
+            && self.buy_amount * other.buy_amount <= self.sell_amount * other.sell_amount
+    }
+}
+
+
+struct Match<'a> {
+    order_pair_type: OrderPairType,
+    orders: OrderPair<'a>,
+}
+
+type OrderPair<'a> = [&'a Order; 2];
+
+type MatchedAmounts = [U256; 2];
+
+
 
 pub fn solve_pair(
     sell_orders_token0: &Vec<Order>,
     sell_orders_token1: &Vec<Order>,
 ) -> Result<Solution> {
+    // Can we assume that the orders are sorted already?
+    assert!(check_orders_sorted_by_limit_price(&sell_orders_token0));
+    assert!(check_orders_sorted_by_limit_price(&sell_orders_token1));
+    // Or do we need to sort them first here?
+    //sell_orders_token0.sort();
+    //sell_orders_token1.sort();
+
+    // Get number of orders in each direction.
+    let nr_orders_token0 = sell_orders_token0.len();
+    let nr_orders_token1 = sell_orders_token1.len();
+
+    // Init vectors of sell amounts.
+    let mut sell_volumes_token0 = vec![U256::zero(); nr_orders_token0];
+    let mut sell_volumes_token1 = vec![U256::zero(); nr_orders_token1];
+
+    // Match orders with best limit prices, if possible.
+    if !(sell_orders_token0.is_empty() || sell_orders_token1.is_empty()) {
+
+        // The best orders are the last elements in their vectors.
+        let best_sell_order_token0 = &sell_orders_token0[nr_orders_token0 - 1];
+        let best_sell_order_token1 = &sell_orders_token1[nr_orders_token1 - 1];
+
+        if let Some(order_pair_type) = best_sell_order_token0.match_compare(
+            &best_sell_order_token1
+        ) {
+            let matched_amounts = get_matched_amounts(
+                &Match {
+                    order_pair_type,
+                    orders: [best_sell_order_token0, best_sell_order_token1],
+                }
+            );
+            sell_volumes_token0[nr_orders_token0 - 1] = matched_amounts[0];
+            sell_volumes_token1[nr_orders_token1 - 1] = matched_amounts[1];
+        };
+    };
+
     return Ok(Solution {
         sell_orders_token0: sell_orders_token0.clone(),
```

Comment:
As we just discussed in person, lets put only the filled orders into the solution vector. Since all orders will be fully filled, our solution struct should not have the sell_volumes_token0 and buy_volumes_token0. I think it would be cool, if you could do it in this PR as well

Reply:
Ok, changed that in 3efc3f4